README.md
# DevSec MySQL Baseline

This Compliance Profile ensures, that all hardening projects keep the same quality.

- <https://github.com/dev-sec/ansible-collection-hardening/tree/master/roles/mysql_hardening>
- <https://github.com/dev-sec/chef-mysql-hardening>
- <https://github.com/dev-sec/puppet-mysql-hardening>

## Standalone Usage

This Compliance Profile requires [cinc-auditor](https://cinc.sh/start/auditor/) or [InSpec](https://github.com/chef/inspec) for execution:

```bash
git clone https://github.com/dev-sec/mysql-baseline
cinc-auditor exec mysql-baseline
```

You can also execute the profile directly from Github:

```bash
cinc-auditor exec https://github.com/dev-sec/mysql-baseline
```

## License and Author

- Author:: Edmund Haselwanter <me@ehaselwanter.com>
- Author:: Dominik Richter <dominik.richter@googlemail.com>
- Author:: Patrick Muench <patrick.muench1111@googlemail.com>
- Author:: Christoph Hartmann <chris@lollyrock.com>

- Copyright 2014, Deutsche Telekom AG
- Copyright 2014-2016, The Hardening Framework Team
- Copyright 2016-2022, DevSec Hardening Framework Team

```text
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# Changelog

## [5.1.0](https://github.com/dev-sec/mysql-baseline/tree/5.1.0) (2023-11-12)

[Full Changelog](https://github.com/dev-sec/mysql-baseline/compare/5.0.0...5.1.0)

**Implemented enhancements:**

- add filter for MySQL roles [\#82](https://github.com/dev-sec/mysql-baseline/pull/82) ([schurzi](https://github.com/schurzi))

**Merged pull requests:**

- add spellchecking with codespell [\#81](https://github.com/dev-sec/mysql-baseline/pull/81) ([schurzi](https://github.com/schurzi))
- Configure Renovate [\#80](https://github.com/dev-sec/mysql-baseline/pull/80) ([renovate[bot]](https://github.com/apps/renovate))

## [5.0.0](https://github.com/dev-sec/mysql-baseline/tree/5.0.0) (2022-12-19)

[Full Changelog](https://github.com/dev-sec/mysql-baseline/compare/4.1.0...5.0.0)

**Implemented enhancements:**

- use centralised issue templates and workflows [\#79](https://github.com/dev-sec/mysql-baseline/pull/79) ([schurzi](https://github.com/schurzi))
- support for el9 [\#76](https://github.com/dev-sec/mysql-baseline/pull/76) ([rndmh3ro](https://github.com/rndmh3ro))

**Merged pull requests:**

- Delete first control [\#78](https://github.com/dev-sec/mysql-baseline/pull/78) ([rndmh3ro](https://github.com/rndmh3ro))
- add suse support [\#77](https://github.com/dev-sec/mysql-baseline/pull/77) ([rndmh3ro](https://github.com/rndmh3ro))

## [4.1.0](https://github.com/dev-sec/mysql-baseline/tree/4.1.0) (2022-08-16)

[Full Changelog](https://github.com/dev-sec/mysql-baseline/compare/4.0.4...4.1.0)

**Merged pull requests:**

- fix wrong process name [\#75](https://github.com/dev-sec/mysql-baseline/pull/75) ([rndmh3ro](https://github.com/rndmh3ro))
- debian 11 has a mysql@localhost user with all privs [\#74](https://github.com/dev-sec/mysql-baseline/pull/74) ([rndmh3ro](https://github.com/rndmh3ro))
- process name and the service name differ on debian 11 [\#73](https://github.com/dev-sec/mysql-baseline/pull/73) ([rndmh3ro](https://github.com/rndmh3ro))
- do not run logfile check if no logfile is specified [\#72](https://github.com/dev-sec/mysql-baseline/pull/72) ([rndmh3ro](https://github.com/rndmh3ro))
- Change linting to Cookstyle [\#71](https://github.com/dev-sec/mysql-baseline/pull/71) ([schurzi](https://github.com/schurzi))

## [4.0.4](https://github.com/dev-sec/mysql-baseline/tree/4.0.4) (2022-01-12)

[Full Changelog](https://github.com/dev-sec/mysql-baseline/compare/4.0.3...4.0.4)

**Implemented enhancements:**

- use service name for process check instead of hardcoding it [\#65](https://github.com/dev-sec/mysql-baseline/pull/65) ([rndmh3ro](https://github.com/rndmh3ro))

**Fixed bugs:**

- fix service process control check [\#67](https://github.com/dev-sec/mysql-baseline/pull/67) ([rndmh3ro](https://github.com/rndmh3ro))

**Closed issues:**

- secure-auth is deprecated in MySQL 8.0.3 [\#58](https://github.com/dev-sec/mysql-baseline/issues/58)

**Merged pull requests:**

- use input instead of attribute [\#70](https://github.com/dev-sec/mysql-baseline/pull/70) ([micheelengronne](https://github.com/micheelengronne))
- format and update README.md [\#69](https://github.com/dev-sec/mysql-baseline/pull/69) ([schurzi](https://github.com/schurzi))
- add dependency to chef-config for CI [\#66](https://github.com/dev-sec/mysql-baseline/pull/66) ([schurzi](https://github.com/schurzi))
- use version tag for changelog action [\#64](https://github.com/dev-sec/mysql-baseline/pull/64) ([schurzi](https://github.com/schurzi))
- Fix lint [\#63](https://github.com/dev-sec/mysql-baseline/pull/63) ([schurzi](https://github.com/schurzi))
- add github action for testing [\#62](https://github.com/dev-sec/mysql-baseline/pull/62) ([rndmh3ro](https://github.com/rndmh3ro))

## [4.0.3](https://github.com/dev-sec/mysql-baseline/tree/4.0.3) (2020-10-20)

[Full Changelog](https://github.com/dev-sec/mysql-baseline/compare/4.0.2...4.0.3)

**Fixed bugs:**

- fix quotes in library again [\#61](https://github.com/dev-sec/mysql-baseline/pull/61) ([rndmh3ro](https://github.com/rndmh3ro))

## [4.0.2](https://github.com/dev-sec/mysql-baseline/tree/4.0.2) (2020-10-20)

[Full Changelog](https://github.com/dev-sec/mysql-baseline/compare/4.0.1...4.0.2)

**Closed issues:**

- Dynamically determine logfile path [\#53](https://github.com/dev-sec/mysql-baseline/issues/53)
- mysql 5.7.6 password -\> authentication\_string [\#35](https://github.com/dev-sec/mysql-baseline/issues/35)

**Merged pull requests:**

- fix mysql\_version command [\#60](https://github.com/dev-sec/mysql-baseline/pull/60) ([rndmh3ro](https://github.com/rndmh3ro))
- use custom resource to get mysql version and distribution [\#59](https://github.com/dev-sec/mysql-baseline/pull/59) ([rndmh3ro](https://github.com/rndmh3ro))
- check if a password column exists and only then check contents [\#57](https://github.com/dev-sec/mysql-baseline/pull/57) ([rndmh3ro](https://github.com/rndmh3ro))

## [4.0.1](https://github.com/dev-sec/mysql-baseline/tree/4.0.1) (2020-08-28)

[Full Changelog](https://github.com/dev-sec/mysql-baseline/compare/4.0.0...4.0.1)

**Merged pull requests:**

- only run password-checks when the appropriate columns exist [\#52](https://github.com/dev-sec/mysql-baseline/pull/52) ([rndmh3ro](https://github.com/rndmh3ro))

## [4.0.0](https://github.com/dev-sec/mysql-baseline/tree/4.0.0) (2020-08-26)

[Full Changelog](https://github.com/dev-sec/mysql-baseline/compare/3.1.7...4.0.0)

**Merged pull requests:**

- BREAKING: config-files should be owned by mysql-user [\#56](https://github.com/dev-sec/mysql-baseline/pull/56) ([rndmh3ro](https://github.com/rndmh3ro))

## [3.1.7](https://github.com/dev-sec/mysql-baseline/tree/3.1.7) (2020-08-17)

[Full Changelog](https://github.com/dev-sec/mysql-baseline/compare/3.1.6...3.1.7)

**Fixed bugs:**

- fix wrong parameter in new dynamic determination [\#55](https://github.com/dev-sec/mysql-baseline/pull/55) ([rndmh3ro](https://github.com/rndmh3ro))


controls/mysql_spec.rb
# encoding: utf-8
#
# Copyright 2014, Deutsche Telekom AG
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

user = attribute('User', description: 'MySQL database user', default: 'root', required: true)
pass = attribute('Password', description: 'MySQL database password', default: 'iloverandompasswordsbutthiswilldo', required: true)

mysql_hardening_file = '/etc/mysql/conf.d/hardening.cnf'

# set OS-dependent filenames and paths
case os[:family]
when 'ubuntu', 'debian'
  mysql_config_path = '/etc/mysql/'
  mysql_config_file = mysql_config_path + 'my.cnf'
  mysql_data_path = '/var/lib/mysql/'
  mysql_log_path = '/var/log/'
  mysql_log_file = 'mysql.log'
  mysql_log_group = 'adm'
  service_name = 'mysql'
when 'redhat', 'fedora'
  mysql_config_path = '/etc/'
  mysql_config_file = mysql_config_path + 'my.cnf'
  mysql_data_path = '/var/lib/mysql/'
  mysql_log_path = '/var/log/'
  mysql_log_file = 'mysqld.log'
  mysql_log_group = 'mysql'
  service_name = 'mysqld'
  service_name = 'mariadb' if os[:release] >= '7'
end

describe service(service_name) do
  it { should be_enabled }
  it { should be_running }
end

# Checking MySQL-databases for risky entries
describe command("mysql -u#{user} -p#{pass} mysql -s -e 'select version();' | tail -1") do
  its(:stdout) { should_not match(/Community/) }
end

describe command("mysql -u#{user} -p#{pass} mysql -s -e 'select substring(version(),1,1);' | tail -1") do
  its(:stdout) { should match(/^5/) }
end

describe command("mysql -u#{user} -p#{pass} -s -e 'show databases like \"test\";'") do
  its(:stdout) { should_not match(/test/) }
end

describe command("mysql -u#{user} -p#{pass} mysql -s -e 'select count(*) from mysql.user where user=\"\";' | tail -1") do
  its(:stdout) { should match(/^0/) }
end

describe command("mysql -u#{user} -p#{pass} mysql -s -e 'select count(*) from mysql.user where length(password)=0 or password=\"\";' | tail -1") do
  its(:stdout) { should match(/^0/) }
end

describe command("mysql -u#{user} -p#{pass} mysql -s -e 'select count(*) from mysql.user where grant_priv=\"y\" and User!=\"root\" and User!=\"debian-sys-maint\";' | tail -1") do
  its(:stdout) { should match(/^0/) }
end

describe command("mysql -u#{user} -p#{pass} mysql -s -e 'select count(*) from mysql.user where host=\"%\"' | tail -1") do
  its(:stdout) { should match(/^0/) }
end

describe command("mysql -u#{user} -p#{pass} mysql -s -e 'select count(*) from mysql.user where user=\"root\" and host not in (\"localhost\",\"127.0.0.1\",\"::1\")' | tail -1") do
  its(:stdout) { should match(/^0/) }
end

# 'Check for multiple instances' do
describe command('ps aux | grep mysqld | egrep -v "grep|mysqld_safe|logger" | wc -l') do
  its(:stdout) { should match(/^1$/) }
end

# Parsing configfiles for unwanted entries
describe mysql_conf.params('mysqld') do
  its('safe-user-create') { should cmp 1 }
  its('old_passwords') { should_not cmp 1 }
  its('secure-auth') { should cmp 1 }
  its('user') { should cmp 'mysql' }
  its('skip-symbolic-links') { should cmp 1 }
  its('secure-file-priv') { should_not eq nil }
  its('local-infile') { should cmp 0 }
  its('skip-show-database') { should eq '' }
  its('skip-grant-tables') { should eq nil }
  its('allow-suspicious-udfs') { should cmp 0 }
end

# binding.pry

# 'Mysql-config: owner, group and permissions'
describe file(mysql_data_path) do
  it { should be_directory }
  it { should be_owned_by 'mysql' }
  it { should be_grouped_into 'mysql' }
end

describe file("#{mysql_data_path}/ibdata1") do
  it { should be_owned_by 'mysql' }
  it { should be_grouped_into 'mysql' }
  it { should_not be_readable.by('others') }
  it { should_not be_writable.by('others') }
  it { should_not be_executable.by('others') }
end

describe file("#{mysql_log_path}/#{mysql_log_file}") do
  it { should be_owned_by 'mysql' }
  it { should be_grouped_into mysql_log_group }
  it { should_not be_readable.by('others') }
  it { should_not be_writable.by('others') }
  it { should_not be_executable.by('others') }
end

describe file(mysql_config_file) do
  it { should be_file }
  it { should be_owned_by 'root' }
  it { should be_grouped_into 'root' }
  it { should_not be_readable.by('others') }
end

# test this only if we have a mysql_hardening_file
if command("ls #{mysql_hardening_file}").exit_status.zero?
  describe file(mysql_hardening_file) do
    it { should be_owned_by 'root' }
    it { should be_grouped_into 'root' }
    it { should_not be_readable.by('others') }
  end
end

# 'Mysql environment'
describe command('env') do
  its(:stdout) { should_not match(/^MYSQL_PWD=/) }
end


default/serverspec/mysql_spec.rb
# encoding: utf-8
#
# Copyright 2014, Deutsche Telekom AG
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'spec_helper'

ENV['mysql_password'] = 'iloverandompasswordsbutthiswilldo'

RSpec::Matchers.define :match_key_value do |key, value|
  match do |actual|
    actual =~ /^\s*?#{key}\s*?=\s*?#{value}/
  end
end

mysql_hardening_file = '/etc/mysql/conf.d/hardening.cnf'

# set OS-dependent filenames and paths
case os[:family]
when 'ubuntu', 'debian'
  mysql_config_file = '/etc/mysql/my.cnf'
  mysql_config_path = '/etc/mysql/'
  mysql_data_path = '/var/lib/mysql/'
  mysql_log_path = '/var/log/'
  mysql_log_file = 'mysql.log'
  mysql_log_group = 'adm'
  os[:release] == '14.04' ? mysql_log_dir_group = 'syslog' : mysql_log_dir_group = 'root'
  service_name = 'mysql'
when 'redhat', 'fedora'
  mysql_config_file = '/etc/my.cnf'
  mysql_config_path = '/etc/'
  mysql_data_path = '/var/lib/mysql/'
  mysql_log_path = '/var/log/'
  mysql_log_file = 'mysqld.log'
  mysql_log_group = 'mysql'
  mysql_log_dir_group = 'root'
  service_name = 'mysqld'
end

tmp_config_file = '/root/tmp-my.cnf'

describe service("#{service_name}") do
  it { should be_enabled }
  it { should be_running }
end

# temporarily combine config-files and remove spaces
describe 'Combining configfiles' do
  describe command("cat #{mysql_config_file} | tr -s [:space:]  > #{tmp_config_file}; cat #{mysql_hardening_file} | tr -s [:space:] >> #{tmp_config_file}") do
    its(:exit_status) { should eq 0 }
  end
end

describe 'Checking MySQL-databases for risky entries' do
  describe command("mysql -uroot -p#{ENV['mysql_password']} mysql -s -e 'select version();' | tail -1") do
    its(:stdout) { should_not match(/Community/) }
  end

  describe command("mysql -uroot -p#{ENV['mysql_password']} mysql -s -e 'select substring(version(),1,1);' | tail -1") do
    its(:stdout) { should match(/^5/) }
  end

  describe command("mysql -uroot -p#{ENV['mysql_password']} -s -e 'show databases like \"test\";'") do
    its(:stdout) { should_not match(/test/) }
  end

  describe command("mysql -uroot -p#{ENV['mysql_password']} mysql -s -e 'select count(*) from mysql.user where user=\"\";' | tail -1") do
    its(:stdout) { should match(/^0/) }
  end

  describe command("mysql -uroot -p#{ENV['mysql_password']} mysql -s -e 'select count(*) from mysql.user where length(password)=0 or password=\"\";' | tail -1") do
    its(:stdout) { should match(/^0/) }
  end

  describe command("mysql -uroot -p#{ENV['mysql_password']} mysql -s -e 'select count(*) from mysql.user where grant_priv=\"y\" and User!=\"root\" and User!=\"debian-sys-maint\";' | tail -1") do
    its(:stdout) { should match(/^0/) }
  end

  describe command("mysql -uroot -p#{ENV['mysql_password']} mysql -s -e 'select count(*) from mysql.user where host=\"%\"' | tail -1") do
    its(:stdout) { should match(/^0/) }
  end

  describe command("mysql -uroot -p#{ENV['mysql_password']} mysql -s -e 'select count(*) from mysql.user where user=\"root\" and host not in (\"localhost\",\"127.0.0.1\",\"::1\")' | tail -1") do
    its(:stdout) { should match(/^0/) }
  end
end

describe 'Check for multiple instances' do
  describe command('ps aux | grep mysqld | egrep -v "grep|mysqld_safe|logger" | wc -l') do
    its(:stdout) { should match(/^1$/) }
  end
end

describe 'Parsing configfiles for unwanted entries' do
  describe file(tmp_config_file) do
    its(:content) { should match_key_value('safe-user-create', '1') }
    its(:content) { should_not match_key_value('old_passwords', '1') }
    its(:content) { should match_key_value('secure-auth', '1') }
    its(:content) { should match_key_value('user', 'mysql') }
    its(:content) { should match_key_value('skip-symbolic-links', '1') }
    its(:content) { should match(/^\s*?secure-file-priv/) }
    its(:content) { should match_key_value('local-infile', '0') }
    its(:content) { should match(/^\s*?skip-show-database/) }
    its(:content) { should_not match(/^\s*?skip-grant-tables/) }
    its(:content) { should match_key_value('allow-suspicious-udfs', '0') }
  end
end

describe 'Mysql-data owner, group and permissions' do
  describe file(mysql_data_path) do
    it { should be_directory }
    it { should be_owned_by 'mysql' }
    it { should be_grouped_into 'mysql' }
  end

  describe file("#{mysql_data_path}/ibdata1") do
    it { should be_owned_by 'mysql' }
    it { should be_grouped_into 'mysql' }
    it { should_not be_readable.by('others') }
    it { should_not be_writable.by('others') }
    it { should_not be_executable.by('others') }
  end

  describe file(mysql_log_path) do
    it { should be_directory }
    it { should be_owned_by 'root' }
    it { should be_grouped_into mysql_log_dir_group }
  end

  describe file("#{mysql_log_path}/#{mysql_log_file}") do
    it { should be_owned_by 'mysql' }
    it { should be_grouped_into mysql_log_group }
    it { should_not be_readable.by('others') }
    it { should_not be_writable.by('others') }
    it { should_not be_executable.by('others') }
  end
end

describe 'Mysql-config: owner, group and permissions' do
  describe file(mysql_config_path) do
    it { should be_directory }
    it { should be_owned_by 'root' }
    it { should be_grouped_into 'root' }
    it { should be_owned_by 'root' }
    it { should be_grouped_into 'root' }
    it { should_not be_readable.by('others') }
  end

  # test this only if we have a mysql_hardening_file
  if command("ls #{mysql_hardening_file}").exit_status == 0
    describe file(mysql_hardening_file) do
      it { should be_owned_by 'mysql' }
      it { should be_grouped_into 'root' }
      it { should_not be_readable.by('others') }
    end
  end
end

describe 'Mysql environment' do
  describe command('env') do
    its(:stdout) { should_not match(/^MYSQL_PWD=/) }
  end
end


controls/mysql_db.rb
# frozen_string_literal: true

# Copyright:: 2014, Deutsche Telekom AG
# Copyright:: 2018, Christoph Hartmann
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

user = input('user')
pass = input('password')

control 'mysql-db-01' do
  impact 0.3
  title 'use supported mysql version in production'
  describe command("mysql -u#{user} -p#{pass} -sN -e 'select version();'") do
    its(:stdout) { should_not match(/Community/) }
  end
end

control 'mysql-db-02' do
  impact 0.5
  title 'use mysql version 5 or higher'
  describe command("mysql -u#{user} -p#{pass} -sN -e 'select substring_index(version(),\".\",1);'") do
    its(:stdout) { should cmp >= 5 }
  end
end

control 'mysql-db-03' do
  impact 1.0
  title 'test database must be deleted'
  describe command("mysql -u#{user} -p#{pass} -sN -e 'show databases like \"test\";'") do
    its(:stdout) { should_not match(/test/) }
  end
end

control 'mysql-db-04' do
  impact 1.0
  title 'deactivate anonymous user names'
  describe command("mysql -u#{user} -p#{pass} -sN -e 'select count(*) from mysql.user where user=\"\";'") do
    its(:stdout) { should match(/^0/) }
  end
end

# current versions of MySQL support roles, we don't check them at the moment.

if command("mysql -u#{user} -p#{pass} -sN -e 'SELECT COUNT(1) FROM information_schema.COLUMNS  WHERE TABLE_SCHEMA = \"mysql\" AND TABLE_NAME = \"user\" AND COLUMN_NAME = \"is_role\"'").stdout.strip == '0'
  role_filter = ''
else
  role_filter = 'AND CONVERT(is_role USING utf8) = "N"'
end

# MySQL 5.7.6 dropped the "password" column in the mysql.user table
# so we have to check if it's there before we check if a password is empty
control 'mysql-db-05' do
  impact 1.0
  title 'default passwords must be changed'
  only_if { command("mysql -u#{user} -p#{pass} -sN -e 'select count(*) from information_schema.columns where table_name=\"user\" and table_schema=\"mysql\" and column_name=\"password\";'").stdout.strip == '1' }
  describe command("mysql -u#{user} -p#{pass} -sN -e 'select count(*) from mysql.user where (length(password)=0 or password=\"\") and (length(authentication_string)=0 or authentication_string=\"\") and not (user=\"mariadb.sys\" and host=\"localhost\") #{role_filter};'") do
    its(:stdout) { should match(/^0/) }
  end
end

# MySQL versions older than 5.7.6 and MariaDB databases still have the
# password column so we need to check if it is empty
control 'mysql-db-05b' do
  impact 1.0
  title 'default passwords must be changed'
  only_if { command("mysql -u#{user} -p#{pass} -sN -e 'select count(*) from information_schema.columns where table_name=\"user\" and table_schema=\"mysql\" and column_name=\"password\";'").stdout.strip == '0' }
  describe command("mysql -u#{user} -p#{pass} -sN -e 'select count(*) from mysql.user where length(authentication_string)=0 or authentication_string=\"\" and not (user=\"mariadb.sys\" and host=\"localhost\") #{role_filter};'") do
    its(:stdout) { should match(/^0/) }
  end
end

control 'mysql-db-06' do
  impact 0.5
  title 'the grant option must not be used'
  describe command("mysql -u#{user} -p#{pass} -sN -e 'select count(*) from mysql.user where grant_priv=\"y\" and user!=\"root\" and user!=\"debian-sys-maint\" and not (user=\"mysql\" and host=\"localhost\");'") do
    its(:stdout) { should match(/^0/) }
  end
end

control 'mysql-db-07' do
  impact 0.5
  title 'ensure no wildcards are used for hostnames'
  describe command("mysql -u#{user} -p#{pass} -sN -e 'select count(*) from mysql.user where host=\"%\"'") do
    its(:stdout) { should match(/^0/) }
  end
end

control 'mysql-db-08' do
  impact 0.5
  title 'it must be ensured that superuser can login via localhost only'
  describe command("mysql -u#{user} -p#{pass} -sN -e 'select count(*) from mysql.user where user=\"root\" and host not in (\"localhost\",\"127.0.0.1\",\"::1\")'") do
    its(:stdout) { should match(/^0/) }
  end
end


libraries/mysql_version.rb
# frozen_string_literal: true

# Copyright:: 2020, Sebastian Gumprich
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# author: Sebastian Gumprich

class MySQLVersion < Inspec.resource(1)
  name 'mysql_version'

  attr_reader :user, :pass

  def initialize(user, pass)
    super()
    @user = user
    @pass = pass
  end

  def mysql_version
    inspec.command("mysql -sN -e 'SHOW VARIABLES WHERE variable_name = \"version\"'").stdout.strip.split("\t")[1].split('-')[0].to_s
  end
end


libraries/mysql_distribution.rb
# frozen_string_literal: true

# Copyright:: 2020, Sebastian Gumprich
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# author: Sebastian Gumprich

class MySQLDistribution < Inspec.resource(1)
  name 'mysql_distribution'

  attr_reader :user, :pass

  def initialize(user, pass)
    super()
    @user = user
    @pass = pass
  end

  def mysql_distribution
    if inspec.command("mysql -u#{user} -p#{pass} -sN -e 'SHOW VARIABLES WHERE Variable_name = \"version_comment\"'").stdout.strip.split("\t")[1].downcase.include? 'mariadb'
      'mariadb'
    else
      'mysql'
    end
  end
end


default/serverspec/spec_helper.rb
# encoding: utf-8
#
# Copyright 2014, Deutsche Telekom AG
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

if ENV['STANDALONE_SPEC']

  require 'serverspec'
  require 'pathname'
  require 'net/ssh'
  require 'highline/import'

  include Serverspec::Helper::Ssh
  include Serverspec::Helper::Exec
  include Serverspec::Helper::DetectOS

  RSpec.configure do |c|

    if ENV['ASK_SUDO_PASSWORD']
      c.sudo_password = ask('Enter sudo password: ') { |q| q.echo = false }
    else
      c.sudo_password = ENV['SUDO_PASSWORD']
    end

    options = {}

    if ENV['ASK_LOGIN_PASSWORD']
      options[:password] = ask("\nEnter login password: ") { |q| q.echo = false }
    else
      options[:password] = ENV['LOGIN_PASSWORD']
    end

    if ENV['ASK_LOGIN_USERNAME']
      user = ask("\nEnter login username: ") { |q| q.echo = false }
    else
      user = ENV['LOGIN_USERNAME'] || ENV['user'] || Etc.getlogin
    end

    if user.nil?
      puts 'specify login user env LOGIN_USERNAME= or user='
      exit 1
    end

    c.host  = ENV['TARGET_HOST']
    options.merge(Net::SSH::Config.for(c.host))
    c.ssh   = Net::SSH.start(c.host, user, options)
    c.os    = backend.check_os

  end

else
  require 'serverspec'
  require 'pathname'

  include Serverspec::Helper::Exec
  include Serverspec::Helper::DetectOS

  RSpec.configure do |c|
    c.before :all do
      c.os = backend(Serverspec::Commands::Base).check_os
    end
  end
end


controls/mysql_conf.rb
# frozen_string_literal: true

# Copyright:: 2014, Deutsche Telekom AG
# Copyright:: 2018, Christoph Hartmann
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

mysql_hardening_file = '/etc/mysql/conf.d/hardening.cnf'

user = input('user')
pass = input('password')

# get mysql version and distribution
version = mysql_version(user, pass)
distribution = mysql_distribution(user, pass)

# get datadir and logfile-path from settings in the configuration if it is defined or from mysql itself
mysql_data_path = if mysql_conf&.params&.mysqld&.datadir
                    mysql_conf.params.mysqld.datadir
                  else
                    command("mysql -u#{user} -p#{pass} -sN -e \"select @@GLOBAL.datadir\";").stdout.strip
                  end

mysql_log_file = if mysql_conf.params.mysqld && mysql_conf.params.mysqld['log-error']
                   mysql_conf.params.mysqld['log-error']
                 else
                   command("mysql -u#{user} -p#{pass} -sN -e \"select @@GLOBAL.log_error\";").stdout.strip
                 end

# set OS-dependent filenames and paths
case os[:family]
when 'ubuntu', 'debian'
  mysql_config_path = '/etc/mysql/'
  mysql_config_file = "#{mysql_config_path}my.cnf"
  mysql_log_group = 'adm'
  process_name = 'mysqld'
  process_name = 'mariadbd' if os[:release] >= '11' && os[:name] == 'debian'
when 'redhat', 'fedora'
  mysql_config_path = '/etc/'
  mysql_config_file = "#{mysql_config_path}my.cnf"
  mysql_log_group = 'mysql'
  process_name = 'mysqld'
  process_name = 'mariadbd' if os[:release] >= '9'
when 'suse'
  mysql_config_path = '/etc/'
  mysql_config_file = "#{mysql_config_path}my.cnf"
  mysql_log_group = 'mysql'
  process_name = 'mysqld'
end

# 'Check for multiple instances' do
control 'mysql-conf-02' do
  impact 0.5
  title 'only one instance of mysql should run on a server'
  describe command("pgrep -x -c #{process_name}") do
    its(:stdout) { should match(/^1$/) }
  end
end

# Parsing configfiles for unwanted entries
control 'mysql-conf-03' do
  impact 0.7
  title 'enable secure configurations for mysql'
  describe mysql_conf.params('mysqld') do
    its('safe-user-create') { should cmp 1 }
    its('old_passwords') { should_not cmp 1 }
    its('user') { should cmp 'mysql' }
    its('secure-file-priv') { should_not eq nil }
    its('local-infile') { should cmp 0 }
    its('skip-symbolic-links') { should cmp 1 }
    its('skip-show-database') { should eq '' }
    its('skip-grant-tables') { should eq nil }
    its('allow-suspicious-udfs') { should cmp 0 }
  end
end

# Parsing configfiles for unwanted entries
control 'mysql-conf-03b' do
  impact 0.7
  title 'enable secure configurations for mysql'
  only_if('MySQL version less than 8.0.3 and not mariadb') do
    (Gem::Version.new(version.mysql_version) <= Gem::Version.new('8.0.3') and distribution.mysql_distribution == 'mysql')
  end
  describe mysql_conf.params('mysqld') do
    its('secure-auth') { should cmp 1 }
  end
end

# 'Mysql-config: owner, group and permissions'
control 'mysql-conf-04' do
  impact 0.7
  title 'ensure the mysql data path is owned by mysql user'
  describe file(mysql_data_path) do
    it { should be_directory }
    it { should be_owned_by 'mysql' }
    it { should be_grouped_into 'mysql' }
  end
end

control 'mysql-conf-05' do
  impact 0.5
  title 'ensure data path is owned by mysql user'
  describe file("#{mysql_data_path}/ibdata1") do
    it { should be_owned_by 'mysql' }
    it { should be_grouped_into 'mysql' }
    it { should_not be_readable.by('others') }
    it { should_not be_writable.by('others') }
    it { should_not be_executable.by('others') }
  end
end

control 'mysql-conf-06' do
  impact 0.5
  title 'ensure log file is owned by mysql user'
  only_if { mysql_log_file != '' }
  describe file(mysql_log_file) do
    it { should be_owned_by 'mysql' }
    it { should be_grouped_into mysql_log_group }
    it { should_not be_readable.by('others') }
    it { should_not be_writable.by('others') }
    it { should_not be_executable.by('others') }
  end
end

control 'mysql-conf-07' do
  impact 0.7
  title 'ensure the mysql config file is owned by user root, group mysql'
  describe file(mysql_config_file) do
    it { should be_file }
    it { should be_owned_by 'root' }
    it { should be_grouped_into 'mysql' }
    it { should_not be_readable.by('others') }
  end
end

# test this only if we have a mysql_hardening_file
control 'mysql-conf-08' do
  impact 0.5
  title 'ensure the mysql hardening config file is owned by user root, group mysql'
  describe file(mysql_hardening_file) do
    it { should be_owned_by 'root' }
    it { should be_grouped_into 'mysql' }
    it { should_not be_readable.by('others') }
  end
  only_if { command("ls #{mysql_hardening_file}").exit_status.zero? }
end

# 'Mysql environment'
control 'mysql-conf-09' do
  impact 1.0
  title 'the use of MYSQL_PWD is insecure and can exploit the password'
  describe command('env') do
    its(:stdout) { should_not match(/^MYSQL_PWD=/) }
  end
end
